compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java

```diff
@@ -3944,8 +3944,9 @@ public void visit(BLangTernaryExpr ternaryExpr) {
             } else if (types.isAssignable(thenType, elseType)) {
                 resultType = elseType;
             } else {
-                dlog.error(ternaryExpr.pos, DiagnosticErrorCode.INCOMPATIBLE_TYPES, thenType, elseType);
-                resultType = symTable.semanticError;
+                resultType = BUnionType.create(null, thenType, elseType);
+                types.setImplicitCastExpr(ternaryExpr.thenExpr, thenType, resultType);
```

Comment:
I missed to comment here earlier. I think we can get rid of these castings. Line No 3948 and 3949